### DISCUSSION TITLE
Detecting/Measuring 'Noise'

### DISCUSSION THREAD

**Discussion by dandspach** [ImageMagick/ImageMagick#3158] [Digital Image Processing]
Suppose I have a folder of multiple images from NOAA weather satellites caputed with an SDR via VHF frequencies.  Some of the satellite passes result in amazing processed images, and others are not so good (see examples below).  Can ImageMagick 'score' these images and possibly help me to eliminate the bad ones?  

A few qualifiers:
- The images are not taken from the same angle, so there is no 'baseline' image
- Images can be color or b/w depending on filters applied and time of day

Any assistance would be appreciated...

![NOAA1820201102-092208 color](https://user-images.githubusercontent.com/12740373/105790430-5ef4f180-5f52-11eb-8644-af98ab4b11b9.png)

![NOAA1520210125-075144 color](https://user-images.githubusercontent.com/12740373/105790221-e0984f80-5f51-11eb-9b29-59ee7946e295.png)

**Comment by snibgo** [comment]
Noise can be defined as "detail at the highest frequency, with an amplitude significantly greater than detail at a lower frequency". This is a bit crude, as it counts coastlines etc as noise. 

For your purpose, we might simplify that definition: "noise is detail at high frequency." We can implement that by finding the difference between the input and a small blur of the input. This is light where we have noise, dark otherwise. The mean of the result gives an overall measure of how noisy the image is. Windows CMD syntax:
```
magick nsesamp1.png ( +clone -blur 0x0.5 ) -compose Difference -composite +write x1.png -format %[fx:mean] info:
```
Your two images results in means:
```
0.0852941
0.0284469
```
So your threshold for "is this image noisy?" could be between those values.

You could refine this by first removing red, yellow and cyan pixels, eg by making them transparent and blurring to fill those gap.

**Reply by dandspach** [reply]
Thank you so very much.  This should get me started.

**Reply by dandspach** [reply]
when i replace the file name, I get the following error:

-bash: syntax error near unexpected token `('

**Reply by fmw42** [reply]
Snibgo's code is for Windows which has different syntax than for Linux and Mac OSX.

I would also suggest converting you input images to grayscale

magick nsesamp1.png -colorspace gray \( +clone -blur 0x0.5 \) -compose Difference -composite +write x1.png -format "%[fx:mean]\n" info:

**Comment by snibgo** [comment]
For bash, escape the parentheses: \\( and \\)

**Reply by dandspach** [reply]
Thank you all for comments.  I'm sorry, but I'm confused.  I should clarify that I am using Linux.  When I use the command:

magick NOAA1920210125-180112l.wv.png -colorspace gray ( +clone -blur 0x0.5 ) -compose Difference -composite +write x1.png -format "%[fx:mean]\n" info:**

I still get the same error.  Can someone edit the command above to what it should look like.  Thanks for your patience...

**Comment by snibgo** [comment]
magick NOAA1920210125-180112l.wv.png -colorspace gray \\( +clone -blur 0x0.5 \\) -compose Difference -composite +write x1.png -format "%[fx:mean]\n" info:

I have escaped the parentheses, and removed your final "**".

**Reply by dandspach** [reply]
Thank you all so very much.  It works!  Thanks again...
